Documentation/teaching/so2/assign4.c-kvm-vmm.rst

```diff
@@ -0,0 +1,332 @@
+
+.. code-block:: c
+
+	typedef struct vm {
+		int sys_fd;
+		int fd;
+		/* Memory of the VM */
+		char *mem;
+	} virtual_machine;
+
+
+The first step in initializing the KVM VM is to interract with the [KVM_API](https://www.kernel.org/doc/html/latest/virt/kvm/api.html]. The KVM API is exposed via ``/dev/kvm``. We will be using ioctl calls to call the API. 
+
+The snippet below shows how one can call ``KVM_GET_API_VERSION`` to get the KVM API Version
+
+.. code-block:: c
+
+	int kvm_fd = open("/dev/kvm", O_RDWR);
+	if (kvm_fd < 0) {
+	    perror("open /dev/kvm");
+	    exit(1);
+	}
+
+	int api_ver = ioctl(kvm_fd, KVM_GET_API_VERSION, 0);
+	if (api_ver < 0) {
+	    perror("KVM_GET_API_VERSION");
+	    exit(1);
+	}
+
+Let us now go briefly through how a VMM initializes a VM. This is only the bare bones, a VMM may do lots of other things during VM initialization.
+
+1. We first use KVM_GET_API_VERSION to check that we are running the expected version of KVM, ``KVM_API_VERSION``. 
+2. We now create the VM using ``KVM_CREATE_VM``. Note that calling ``KVM_CREATE_VM`` returns a file descriptor. We will be using this file descriptor for the next phases of the setup. 
+3. (Optional) On Intel based CPUs we will have to call ``KVM_SET_TSS_ADDR`` with address ``0xfffbd000``
+4. Next, we allocate the memory for the VM, we will be using ``mmap`` for this with ``PROT_WRITE``, ``MAP_PRIVATE``, ``MAP_ANONYMOUS`` and ``MAP_NORESERVE``. We recommend allocating 0x100000 bytes for the VM.
+5. We flag the memory as ``MADV_MERGEABLE`` using ``madvise``
+6. Finally, we use ``KVM_SET_USER_MEMORY_REGION`` to assign the memory to the VM.
+
+**Make sure you understand what file descriptor to use and when, we use the KVM fd when calling KVM_CREATE_VM, but when interacting with the vm such as calling KVM_SET_USER_MEMORY_REGION we use the VMs
+file descriptor** 
+
+TLDR: API used for VM initialization:
+
+* KVM_GET_API_VERSION
+* KVM_CREATE_VM
+* KVM_SET_TSS_ADDR
+* KVM_SET_USER_MEMORY_REGION.
+
+2. Initialize a virtual CPU
+___________________________
+
+We need a Virtual CPU (VCPU) to store registers.
+
+.. code-block:: c
+
+	typedef struct vcpu {
+		int fd;
+		struct kvm_run *kvm_run;
+	} virtual_cpu;
+
+To create a virtual CPU we will do the following:
+1. Call ``KVM_CREATE_VCPU`` to create the virtual CPU. This call returns a file descriptor.
+2. Use ``KVM_GET_VCPU_MMAP_SIZE`` to get the size of the shared memory
+3. Allocated the necessary VCPU mem size with ``mmap``. We will be passing the VCPU file descriptor to the ``mmap`` call. We can store the result in ``kvm_run``.
+
+
+TLDR: API used for VM
+
+* KVM_CREATE_VCPU
+* KVM_GET_VCPU_MMAP_SIZE
+
+**We recommend using 2MB pages to simplify the translation process**
+
+Running the VM
+==============
+
+
+Setup real mode
+---------------
+
+At first, the CPU will start in Protected mode. To do run any meaningful code, we will switch the CPU to [Real mode](https://wiki.osdev.org/Real_Mode). To do this we will
```

Comment:
Neither is this one even thought is correct markdown syntax

Reply:
Will send a force push to fix this.